### PR TITLE
Remove event list

### LINKED ISSUE #1281: Rethinking event data
### 1.1 API UX problems

1. The [API for creating event workspaces](https://scipp.github.io/event-data/overview.html) is not very natural. In principle this would be used little in practice since event data would just be read from a file.
2. Requires a new `dtype` for every type of event list, e.g., `sc.dtype.event_list_float64`, `sc.dtype.event_list_int32`.
3. Coords and data of event lists are separate. Slicing a data array down to a single bin will *not* provide access to an event list as a simple data array view --- it is a data array containing lists.
4. [Realignment mechanism](https://scipp.github.io/user-guide/unaligned-data.html) seems to work, but also feels potentially complicated and non-trivial to understand.
5. Realignment mechanism appears to limited when attempting to support the equivalent to Mantid's `MDEventWorkspace`. The API is in principle available, but without an optimization such as the recursive bucketing used by `MDEventWorkspace` slicing and other lookup will be too slow in practice since in the current implementation it involves a full filter operation on a large event list.
6. Realignment mechanism did not naturally offer the expected benefits when it comes to unit conversion: The idea was to avoid the need of converting a 1D coord into a 2D coord by simply having realigning bin edges that could be kept 1D. In practice, the conversion would still create a 2D coord (which fails, because we do not support realignment with a 2D coord), and a [special `realign` option](https://scipp.github.io/generated/scipp.neutron.convert.html#scipp.neutron.convert) had to be added.
7. No mechanism to prevent length-mismatch between, e.g., event weights and event coords.

### 1.2 Other problems

1. For I/O purposes, storing variable-length data is likely inefficient. Copy into contiguous chunk of memory would be a solution, but comes with extra complexity and copy overhead.
2. Hypothetical overhead from copying many small allocated regions instead of a single contiguous chunk. We have not seen large effects of this in practice yet, but may a closer look would be useful.
3. `sc.realign` comes internally in two flavors, one for event data, one for dense data. There are limitations in the implementation, such as not supporting multi-dimensional histogramming for event data yet.

## 2 Wishes for a better solution

1. `array['spectrum', 17].value` should return a dense `DataArrayView` depending on a single dimension such as `'event'`.
2. Underlying data strored in a single large chunk of memory instead of using "lists of lists".
3. Traditional bucketing as used for TOF event lists of raw data to use a mechanism that can also solve the `MDEventWorkspace` lookup/filter performance problem. In particular the current slicing of buckets (`array['spectrum', 17]`) should be combinable with a more precise selection:
   ```python
   # bucket edges
   qx = array.coords['Qx']
   # qx is Variable(dims=['Qx'], shape=[100], dtype=float64)

   # event Qx
   qx = array['Qx', 4]['Qy', 5]['Qz',6].value.coords['Qx']
   # qx is Variable(dims=['event'], shape=[2342], dtype=float32)
   # note how value in bucket has independent coord
   ```
4. Realignment mechanism made redundant?

## 3 Possible solution

There may be a surprisingly simple solution based on a specialization of [`DataModel`](https://scipp.github.io/developer/variable_implementation.html):
```cpp
template <>
class DataModel<event_list> {
  element_array<pair<index, index>> m_buckets; // list of slice begin/end
  Dim m_dim; // dim to slice
  DataArray m_events;
};
```

While there are *a lot* of unclear details and a number of technical implementation challenges (see below), we can consider how this solves problems and implements wishes discussed above.

- 2.2 (and thus 1.2.1 and 1.2.2) solved by contiguous `m_events`.
- 1.1.2 replaced by a single new `sc.dtype.event_list` (better, more general name may be needed)
- 2.1 solved since we can return a slice of `m_events`.
- 2.3 should just work, since `m_dim` in the `DataModel` specialization is independent of the dims of the underlying `VariableConcept`, i.e., buckets can form a grid of arbitrary shape.
- 1.1.3 solved since an event list would just be a `DataArray`. Access to an event list within a data array would be via `DataArrayView`, i.e., without option to modify size.
- As a surprising side-benefit, this might solve 80% of #1211, by providing a slicable variable that contains a single underlying data array. As a corollary, we should also support `Dataset` here, e.g., by actually specializing as `template <class T> class DataModel<bucket<T>>` where `T` could be `DataArray` or `Dataset`.
- 1.2.3 may be solved, since bucketing is not specific to event data. Can accept and afford reordering/sorting data? Probably sufficient to reorder buckets in case of realigning, e.g., both events and positions. Mechanism simpler since just nested variable.
- 1.1.6 is *not* solved in general. However, the bucket mechanism gives us at least the option to naturally add and subtract event data, without the need to add bin edges for the TOF dimension. That is, we may be able to reduce the number of cases where we run into trouble with 2d coords from unit conversion. For a full solution we would need some  sort of support for dynamic buckets (see inconclusive discussion below).
- 1.1.7 solved naturally by `DataArray` shape checks.

### Limitations

1. Resizing individual event lists not readily possible without "list of lists" data structure. In practice resizing occurs relatively little. If it turns out to be more common that though, we *could* implement an over-allocation mechanism similar to what `std::vector` provides: `m_events` should contain padding between buckets, so individual buckets could be grown until hitting the following bucket, at which point reallacation would become necessary. It is unclear whether this is technically feasible though, since it would not provide multi-threaded size-modifying access to different buckets (unlike the current solution).
2. The current realignment mechanism maps coords of the realigned wrapper into the unaligned data. For example, considering the equivalent of Mantid's `EventWorkspace`, a position coord would be mapped from the realigned wrapper (with TOF dimension and edges) into the unaligned array containing TOF event lists, since the position (spectrum) dimension is not affected by the realignment. It does not appear feasible to do something equivalent in the proposed solution, at least not without adding specialized code to `DataArrayView::value()`.

### Open conceptual questions

- We may want to have a variable containing event list with dimensions not present in the bucketing.
  - For example, we may want to overlay a 1000x1000x1000 X-Y-Z grid to data depending on a single position dimension. Creating this many buckets comes with substantial overhead and may even exceed available memory.
  - Another example would be TOF event lists where we may want to add a pulse-time dimension, e.g., for [sample-environment-parameter filtering ](https://scipp.github.io/event-data/filtering.html) and [arithmetic operations with event data](https://scipp.github.io/event-data/overview.html#Arithmetic-operations). The current realignment mechanism supports this without the need of actual sorting/bucketing and does this on-the-fly instead.
  - Maybe most relevant and critical example: TOF event data for a large powder diffractometer. May have more than `1 Mpixel` and more than `1 kbin` required for normalization to monitor. [Arithmetic support between event data and histograms](https://scipp.github.io/event-data/overview.html#Arithmetic-operations) relies on matching binning for event data on realigned wrapper. If we want to carry over this (desirable) requirement we either have to deal with a excessively large number of buckets, or have a way to avoid buckets while still having the related dimension/shape as part of the variable.
  - Should this only be supported using an explicit syntax, i.e., a dedicated function instead of an operator? Should we keep handling it as a special case of `*` for data arrays, even when not "realigned"?

### Open technical implementation problems

- How should low-level access work, in particular for `transform` implementation. Current mechanism of returning vectors from kernels which are then moved into output event lists would not work. Naive implementation would require a copy, so a better approach is needed.
  - Note that in principle it would be sufficient to provide basic operators between POD types and `DataArray` so `transform` would apply kernels to, e.g., `(DataArrayView, double)`. However, this implementation would also be very inefficient (investigation required on how inefficient).
  - More likely, we may need to reconsider [`VariableConceptT<T>`](https://scipp.github.io/developer/variable_implementation.html) to support a third subclass in addition to `DataModel<T>` and `ViewModel<T>`. The latter is a strided view, whereas the new `DataModel` specialization would require and indexed view. We may thus have `DataModel<T>`, `StrideViewModel<T>`, and `IndexViewModel<T>`. This may actually provide additional benefits at no extra cost, see also #454. For example, it could be used to implement value-based slicing with 2D "ragged" coordinates.
  - After closer consideration, it appears that `ViewModel` is entirely unnecessary and we can implement `VariableView` to instead create an `ElementArrayView` on the fly. This would simplify the overall class structure and avoid memory allocation in `VariableView` (when slicing). This improvement is independent of the other changes discussed here and could be done as a preparation step.
  - For accessing element of event lists, we would either need an alternative to `ElementArrayView` (and the underlying `ViewIndex`) that can be used in transform explicitly (similar to old distinction between `span` and `ElementArrayView` between `Variable` and `VariableView`), or, more desirably, adapt `ElementArrayView` to be able to deal with indices obtained from subspans. The latter is possible and certainly more desirable since it could significantly reduce the amount of instantiations required by `transform`. However, iteration of `ElementArrayView` lies at the core of all `tranform`-based functionality and is directly responsible for the performance of most scipp functions. It will need to be tested if a generalized implementation can yield equivalent performance.
- Current module split may not work any more: Need `DataArray` for instantiating, e.g., `operator*(Variable, Variable)`, where one of the variables contains event lists.
  - Iteration of elements can be based on `ElementArrayView` (or similar new views) and is thus generic.
  - Creating output variable with buckets of `DataArray` may be solvable with some part of dynamic factory / copy construction mechanism.
  - Not clear how to handle `+` and `-` between event data, this cannot really be done without including/linking to `DataArray`.
- `Variable::values<T>` returns a `span<T>` (or `ElementArrayView<T>` for `VariableView::values<T>`). We cannot use either of these, since the proposed specialization of `DataModel` does not actually contain a range of `DataArrayView`. Rather, we would need to generate these on-the-fly. This may require specializations of methods of `Variable` and related classes.
  - Note that there are two cases here, access to the nested elements, i.e., individual events, which might be handled using a generalized `ElementArrayView`. To access the "event lists" we would indeed need on-the-fly generation.

## Subtasks

- [x] Support + and - (or equivalent named functions) with buckets (replacing what is supported for realigned data). See #1331.
- [x] Support * and / (or equivalent named functions) between buckets and histograms (replacing what is supported for realigned data). See #1337.
- [x] Support histogramming. See #1334.
- [ ] Support "event filtering" (replacing what is supported for realigned data).
- [x] Support neutron::convert (replacing what is supported for realigned data). See #1341.
- [x] Refactor "realign" docs into "bucket" docs.
- [ ] Improve performance issues introduced in these refactoring steps. We may need to add special case handling in transform, e.g., for stride 1, so the compiler can generate better code. It is currently unclear whether this is an issue in practice. There is a really significant slowdown in the single-threaded case, but we may still be good enough when using threads to be memory-bandwidth limited.
- [x] Remove `event_list`
- [x] Remove `realign`. See #1337.
- [x] Add `to_buckets` that accepts a dict of bin edges instead of indices. This would then sort the buffer (recursively) and determine indices. Similarely, splitting existing buckets by setting new bin edges (for existing or new dims) should be supportable in the same code.
- [x] Replacements for `flatten()` and `groupby.flatten()`, concatenating all buckets along a dimension. Note that we could now in principle do this in-place without reallocating memory, but in practice an in-place reordering of buckets may be too complex.
- [x] All refactor tests and docs.

### LINKED ISSUE #1335: Mantid EventWorkspace to bucket conversion
Currently Mantid's `EventWorkspace` gets converted to variables with, e.g., `dtype=event_list_double`. With #1281 event lists will be removed and replaced by bucket variables. While support for this is not complete yet we can already implement the converter (while keeping the old one in parallel).

- Concatenate all event lists into a single data array.
- Use `to_buckets` to create a bucket variable from this.
  - We may initially add just a single large bucket along the "tof" dimension, i.e., `dims=['spectrum', 'tof']` and `shape=[nspec, 1]`. Ultimately we may want to split buckets along `tof`. This requires sorting with buckets and finding the indices. There are no tools for this implemented yet, so the suggestion would be to postpone this.
- The bucket variable will be the "data" of the data array created from the workspace. Remaining coords and masks are added as usual.
  - The bin edges for `tof` would only use the first and last edge, as long as we support only a single bucket along `tof`.
- For now, the simplest solution may be to add a flag to the converter, to chose how event data gets converted. Later, we remove the old conversion and the flag.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Removes `event_list`, refactor many tests, and remove outdated ones.
- Fixes #1335.
- Add `groupby.concatenate`, replacing `groupby.flatten`.
- Move around and add some helper code the facilitate implementation of above.
- Refactor neutron diffraction tutorial.
- Fixes #1281 (mostly).

Note the growing to-do list in #1363. We are not done yet with this PR, but it is a major step.

### CODE REVIEW

File: python/src/scipp/neutron/instrument_view.py [hunk 1]
```diff
@@ -24,15 +22,13 @@ def _rebin_histogram_data(obj, bins):
...
-        return obj, None
-    else:
-        return obj, new_bins
+    for dim, val in new_bins.items():
+        obj = sc.rebin(obj, dim, su.make_bins(data_array=obj,
+                                              bins=val,
+                                              dim=dim))
+    # Once data has been histogrammed, we return None as the new bins,
+    # since the plot function expects event data if bins are specified.
+    return obj, None
```
Comment (SimonHeybrock) [inline python/src/scipp/neutron/instrument_view.py:31]: @nvaytet Is this the right approach, just remove the old events branch everywhere in the plotting code? I think we will anyway need to make additions so plotting can histogram bucketed data (as a follow-up step)?

File: python/src/scipp/plot/plot.py [hunk 2]
```diff
@@ -83,14 +83,7 @@ def plot(scipp_obj,
...
     tobeplotted = dict()
     for name, var in sorted(inventory.items()):
-
-        if sc.contains_events(var) and bins is None:
-            raise RuntimeError("The `bins` argument must be specified when "
-                               "plotting event data.")
-
         ndims = len(var.dims)
-        if bins is not None and sc.contains_events(var):
-            ndims += 1
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot.py:93]: @nvaytet See also here. I suppose we will eventually want special handling for buckets. Is it fine to just leave it out in this PR?

Comment (nvaytet) [reply python/src/scipp/plot/plot.py:150]: I would say that if you can get the plot to show up and it works as you want, then it's fine.
We will definitely need to add handling of buckets in the plots, so fine to leave it out for now.

File: python/src/scipp/neutron/instrument_view.py (hunk 1 above)
Comment (nvaytet) [reply python/src/scipp/neutron/instrument_view.py:31]: Yes fine for now, see below

File: neutron/test/convert_test.cpp [hunk 3]
```diff
@@ -143,78 +123,44 @@ TEST_P(ConvertTest, Tof_to_DSpacing) {
   const double lambda_to_d = 1.0 / (2.0 * sin(0.5 * asin(0.1 / (L - 10.0))));
 
   ASSERT_TRUE(dspacing.contains("counts") || dspacing.contains("events"));
```
Comment (nvaytet) [inline neutron/test/convert_test.cpp:125]: Is the `"events"` entry still relevant?

File: neutron/test/convert_test.cpp [hunk 4]
```diff
@@ -143,78 +123,44 @@ TEST_P(ConvertTest, Tof_to_DSpacing) {
...
+  EXPECT_EQ(coord.unit(), units::angstrom);
+
+  const auto values = coord.values<double>();
+  // Rule of thumb (https://www.psi.ch/niag/neutron-physics):
+  // v [m/s] = 3956 / \lambda [ Angstrom ]
+  Variable tof_in_seconds = tof.coords()[Dim::Tof] * (1e-6 * units::one);
+  const auto tofs = tof_in_seconds.values<double>();
+  // Spectrum 0 is 11 m from source
+  // 2d sin(theta) = n \lambda
+  // theta = 45 deg => d = lambda / (2 * 1 / sqrt(2))
```
Comment (nvaytet) [inline neutron/test/convert_test.cpp:139]: Define a common factor once?
e.g.
```
fact = 3956.0 / (11.0 * sqrt(2.0))
EXPECT_NEAR(values[0], fact * tofs[0], values[0] * 1e-3);
```

File: python/element_array_view.cpp [hunk 5]
```diff
@@ -60,20 +53,13 @@ void declare_ElementArrayView(py::module &m, const std::string &suffix) {
...
-                               [[maybe_unused]] const T value) {
-      if constexpr (is_view_v<decltype(self[i])>)
-        throw std::runtime_error("Setting items of this type is not possible.");
-      else
-        self[i] = value;
-    });
+  view.def("__setitem__", [](ElementArrayView<T> &self, const scipp::index i,
+                             [[maybe_unused]] const T value) {
+    if constexpr (is_view_v<decltype(self[i])>)
+      throw std::runtime_error("Setting items of this type is not possible.");
```
Comment (nvaytet) [inline python/element_array_view.cpp:59]: Maybe a better message saying that you cannot modify a view?

File: variable/include/scipp/variable/bucket_variable.tcc [hunk 6]
```diff
@@ -27,18 +27,20 @@ VariableConstView::constituents() const {
 }
 
 template <class T>
-std::tuple<VariableConstView, Dim, typename T::element_type>
+std::tuple<VariableView, Dim, typename T::element_type>
```
Comment (nvaytet) [inline variable/include/scipp/variable/bucket_variable.tcc:30]: Can you explain why this change?

File: variable/include/scipp/variable/buckets.h [hunk 7]
```diff
@@ -19,4 +19,8 @@ SCIPP_VARIABLE_EXPORT void copy_slices(const VariableConstView &src,
 [[nodiscard]] SCIPP_VARIABLE_EXPORT std::tuple<Variable, scipp::index>
 sizes_to_begin(const VariableConstView &sizes);
 
+[[nodiscard]] SCIPP_VARIABLE_EXPORT Variable from_constituents(Variable indices,
```
Comment (nvaytet) [inline variable/include/scipp/variable/buckets.h:22]: So passing by value is intentional here? I think i remember seeing this function's body using `move` somewhere above, but I am not sure anymore.

**Review by nvaytet — COMMENTED** [review]
transform looks much nicer :+1: 

This seems like also a good step in making support for event data lees invasive than the old event_list, in case we want to change it again in the future.

Changes to python look fine (including the plotting and plotting tests)

File: dataset/bucket.cpp [hunk 8]
```diff
@@ -267,8 +296,9 @@ Variable histogram(const VariableConstView &data,
     spans = VariableConstView(merged);
   }
   return variable::transform_subspan<std::tuple<
-      args<double, double, double, double>, args<double, float, double, double>,
-      args<double, float, double, float>, args<double, double, float, double>>>(
+      args<float, double, float, double>, args<double, double, double, double>,
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply dataset/bucket.cpp:299]: Not really, especially since here we may want to limit the combinations, e.g., to require that edges and event coord values have the same type, to avoid compiling too many cases.

File: dataset/test/dataset_arithmetic_test.cpp [hunk 9]
```diff
@@ -348,49 +348,6 @@ TYPED_TEST(DatasetBinaryEqualsOpTest, rhs_DatasetView_coord_mismatch) {
                except::CoordMismatchError);
 }
 
-TYPED_TEST(DatasetBinaryEqualsOpTest,
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply dataset/test/dataset_arithmetic_test.cpp:351]: No we do not. I *think* I have equivalents for all the *relevant* ones, but there one you comment on here is actually an example of one that is removed: I believe it is not necessary with the new style of event data. Since it is handled entirely based on the `Variable` level there should be no relevant thing to test for binary ops of `Dataset` and `DataArray` (provided that the interface of `Variable` is not broken). This is an example of how the new event data support reduces complexity of `Dataset`.

File: neutron/test/convert_test.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply neutron/test/convert_test.cpp:139]: Unrelated to this PR, so I suggest keeping it.

File: variable/include/scipp/variable/bucket_variable.tcc (hunk 6 above)
Comment (SimonHeybrock) [reply variable/include/scipp/variable/bucket_variable.tcc:30]: This now supports changing the bucket indices, i.e., the size of each bucket/bin. I am not sure letting clients to this is a good idea (since it is unsafe if out-of-range or overlapping indices are set), but it is also required to grow buckets, e.g., after reserving space and merging bins, see `concatenate` (to be renamed to `merge_bins` in a future PR). Since it is not exposed to Python I am at this point not too worried about the possibility to setup broken indices, but it is certainly something to keep in mind for the future, when we iterate and create a more solid interface for the bucket mechanism.

File: variable/include/scipp/variable/buckets.h (hunk 7 above)
Comment (SimonHeybrock) [reply variable/include/scipp/variable/buckets.h:22]: Yes, passing by value gives us both, copy and move support:

```cpp
Variable var;
from_constituents(var); // makes a copy
from_constituents(std::move(var)); // moves, no copy
```

Passing by value is a way to avoid overloading on `&&`. The downside is that we have to move *twice*, but in practice this does not matter for large objects (it would if designing a class that accepts, e.g., small strings).

File: benchmark/events_histogram_op_benchmark.cpp [hunk 10]
```diff
@@ -12,6 +12,7 @@
 
 using namespace scipp;
 
+/*
```
Comment (Matthew-Andrew) [inline benchmark/events_histogram_op_benchmark.cpp:15]: I assume there is a reason but why is this whole file commented out?

Comment (SimonHeybrock) [reply benchmark/events_histogram_op_benchmark.cpp:15]: Yes, burnout/lazyness. I am adding this to the to-do in #1363, thanks for pointing it out.

File: dataset/include/scipp/dataset/groupby.h [hunk 11]
```diff
@@ -17,7 +17,7 @@ namespace scipp::dataset {
 class SCIPP_DATASET_EXPORT GroupByGrouping {
 public:
   using group = boost::container::small_vector<Slice, 4>;
-  GroupByGrouping(Variable &&key, std::vector<group> &&groups)
+  GroupByGrouping(Variable key, std::vector<group> groups)
```
Comment (Matthew-Andrew) [inline dataset/include/scipp/dataset/groupby.h:20]: Why do we no longer pass by reference here?

Comment (SimonHeybrock) [reply dataset/include/scipp/dataset/groupby.h:20]: To cover both cases: Passing a temporary (moved) as well as making a copy, see https://github.com/scipp/scipp/pull/1359#discussion_r516647717.

File: dataset/bucket.cpp [hunk 12]
```diff
@@ -114,6 +116,58 @@ Dataset resize_default_init(const DatasetConstView &parent, const Dim dim,
...
+  else if (var.dtype() == dtype<bucket<DataArray>>)
+    return bucket_sizes_impl<DataArray>(var);
+  else if (var.dtype() == dtype<bucket<Dataset>>)
+    return bucket_sizes_impl<Dataset>(var);
+  else
+    return makeVariable<scipp::index>(var.dims());
+}
+
+DataArray bucket_sizes(const DataArrayConstView &array) {
+  return {bucket_sizes(array.data()), array.aligned_coords(), array.masks(),
```
Comment (Matthew-Andrew) [inline dataset/bucket.cpp:154]: I'm struggling to understand here why we pass back the unmodified coords and masks along with the bucket sizes. In this case are the aligned coords even guaranteed to be the same length as the new bucket sizes variable?

Comment (SimonHeybrock) [reply dataset/bucket.cpp:154]: Yes, think of this just like any other element-wise unary operation, e.g., `sin()`. The operation is simply applied to each data element, and a new data arrays is returned. In the case of `sin` we transform `array<double> -> array<double>`. `bucket_sizes` transforms `array<DataArrayView> -> array<double>`.

File: python/src/scipp/show.py [hunk 13]
```diff
@@ -114,8 +110,7 @@ def _events_extent(self):
             # but may have not data
             # Find a events coord to use for determining length
             for coord in self._variable.coords.values():
-                if sc.contains_events(coord):
-                    data = coord.values
+                pass
```
Comment (Matthew-Andrew) [inline python/src/scipp/show.py:113]: If we are just passing don't need this loop at all. Can do so in the context above.

File: dataset/bucket.cpp (hunk 12 above)
Comment (Matthew-Andrew) [reply dataset/bucket.cpp:154]: Thanks that makes sense now.

File: dataset/include/scipp/dataset/bucket.h [hunk 14]
```diff
@@ -22,6 +23,25 @@ SCIPP_DATASET_EXPORT void copy_slices(const DatasetConstView &src,
 [[nodiscard]] SCIPP_DATASET_EXPORT Dataset resize_default_init(
     const DatasetConstView &parent, const Dim dim, const scipp::index size);
 
+[[nodiscard]] SCIPP_DATASET_EXPORT Variable from_constituents(Variable indices,
```
Comment (OwenArnold) [inline dataset/include/scipp/dataset/bucket.h:26]: Newly added functions are quite self-explanatory IMHO, but `resize_default_init` could use some comments to describe what it's doing and what the parameters are for.

File: dataset/test/dataset_arithmetic_test.cpp [hunk 15]
```diff
@@ -953,13 +769,19 @@ TEST(DatasetSetData, labels) {
...
-  set_events_variances<double>(bad, {{0.5, 0.6}, {0.8}});
+  Variable indicesGood = makeVariable<std::pair<scipp::index, scipp::index>>(
+      Dims{Dim::X}, Shape{2}, Values{std::pair{0, 2}, std::pair{2, 3}});
+  Variable indicesBad = makeVariable<std::pair<scipp::index, scipp::index>>(
+      Dims{Dim::X}, Shape{2}, Values{std::pair{0, 2}, std::pair{2, 4}});
+  Variable table =
+      makeVariable<double>(Dims{Dim::Event}, Shape{4}, units::m,
+                           Values{1, 2, 3, 4}, Variances{5, 6, 7, 8});
+  Variable good = from_constituents(indicesGood, Dim::Event, table);
+  Variable bad = from_constituents(indicesBad, Dim::Event, table);
```
Comment (OwenArnold) [inline dataset/test/dataset_arithmetic_test.cpp:780]: Why allow even to get this far when the bad indices are out of range?

**Review by OwenArnold — COMMENTED** [review]
I still need to test the updated notebooks. The code looks neat and I couldn't identify any major problems from visual inspection.

Aside from comment about developer docs - As the bucketed code has been reviewed by a number of developers, perhaps without any single one following all PRs in high detail. Maybe we should conclude the removal of the event_lists with a "from first principles" run through of buckets for the development team to ensure everyone knows how to work with this going forward.

File: docs/scipp-neutron/groupby.ipynb [hunk 16]
```diff
@@ -165,7 +165,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "We use `scipp.groupby` with the desired bins and apply a `flatten` operation on dimension `spectrum`.\n",
+    "We use `scipp.groupby` with the desired bins and apply a `concatenate` operation on dimension `spectrum`.\n",
     "This is the events-data equivalent to summing histograms:"
```
Comment (OwenArnold) [inline docs/scipp-neutron/groupby.ipynb:169]: Existing typo "This is the events-data equivalent " -> "This is the event data equivalent " can be dealt with in follow up.

File: dataset/bucket.cpp [hunk 17]
```diff
@@ -114,6 +116,58 @@ Dataset resize_default_init(const DatasetConstView &parent, const Dim dim,
...
+Variable from_constituents(Variable indices, const Dim dim, DataArray buffer) {
+  return from_constituents_impl(std::move(indices), dim, std::move(buffer));
+}
+
+Variable from_constituents(Variable indices, const Dim dim, Dataset buffer) {
+  return from_constituents_impl(std::move(indices), dim, std::move(buffer));
+}
+
+namespace {
+template <class T> Variable bucket_sizes_impl(const VariableConstView &view) {
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply dataset/bucket.cpp:135]: 👍 Added as todo in #1363. I hope to somewhat cleanup the toolset first, avoiding refactoring of tests multiple times.

File: docs/python-reference/api.rst [hunk 18]
```diff
@@ -125,14 +125,12 @@ Reduction
...
    GroupByDataArray.max
    GroupByDataArray.mean
    GroupByDataArray.min
    GroupByDataArray.sum
    GroupByDataset.all
    GroupByDataset.any
-   GroupByDataset.flatten
    GroupByDataset.max
    GroupByDataset.mean
    GroupByDataset.min
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply docs/python-reference/api.rst:136]: Agree, doing so no would be a waste of time, since a lot of these things will get refactored or renamed. Added as todo in #1363.

File: dataset/include/scipp/dataset/bucket.h (hunk 14 above)
Comment (SimonHeybrock) [reply dataset/include/scipp/dataset/bucket.h:26]: Added to doxygen in `.cpp`. Note that the function will likely get renamed, I wrote the doxygen following the future rename buckets->bins.

File: dataset/test/dataset_arithmetic_test.cpp (hunk 15 above)
Comment (SimonHeybrock) [reply dataset/test/dataset_arithmetic_test.cpp:780]: The indices are *not* out of range (`4` is "end", not "last"). They are bad only in the sense that they do not match the others.

**Comment by SimonHeybrock** [comment]
> Maybe we should conclude the removal of the event_lists with a "from first principles" run through of buckets for the development team to ensure everyone knows how to work with this going forward.

Very good idea, though I suggest to wait a couple more steps since not just the public API but also internals needs a lot of consolidation, cleanup, and refactoring.

**Review by OwenArnold — APPROVED** [review]
Good step forward. All my minor concerns are being dealt with as follow ups.

**Comment by SimonHeybrock** [comment]
Windows CI passed previously, and last commit changes only docs -> merging.
